state-sync/storage-service/server/src/lib.rs

```diff
@@ -166,25 +373,144 @@ impl<T: StorageReaderInterface> StorageServiceServer<T> {
+/// Returns the list of peers that made those subscriptions
+/// alongside the ledger info at the target version for the peer.
+fn get_peers_with_ready_subscriptions<T: StorageReaderInterface>(
+    cached_storage_server_summary: Arc<RwLock<StorageServerSummary>>,
+    data_subscriptions: Arc<Mutex<HashMap<AccountAddress, DataSubscriptionRequest>>>,
+    storage: T,
+) -> Result<Vec<(AccountAddress, LedgerInfoWithSignatures)>, Error> {
+    // Fetch the latest storage summary and highest synced version
+    let latest_storage_summary = cached_storage_server_summary.read().clone();
+    let highest_synced_ledger_info = match latest_storage_summary.data_summary.synced_ledger_info {
+        Some(ledger_info) => ledger_info,
+        None => return Ok(vec![]),
+    };
+    let highest_synced_version = highest_synced_ledger_info.ledger_info().version();
+    let highest_synced_epoch = highest_synced_ledger_info.ledger_info().epoch();
+
+    // Identify the peers with ready subscriptions
+    let mut ready_subscriptions = vec![];
+    for (peer, data_subscription) in data_subscriptions.lock().iter() {
+        if data_subscription.highest_known_version() < highest_synced_version {
+            let highest_known_epoch = data_subscription.highest_known_epoch();
+            let target_ledger_info = if highest_known_epoch < highest_synced_epoch {
+                // The peer needs to sync to their epoch ending ledger info
+                get_epoch_ending_ledger_info(highest_known_epoch, storage.clone())?
+            } else {
+                highest_synced_ledger_info.clone()
+            };
+            ready_subscriptions.push((*peer, target_ledger_info));
+        }
+    }
+    Ok(ready_subscriptions)
+}
+
+/// Gets the epoch ending ledger info at the given epoch
+fn get_epoch_ending_ledger_info<T: StorageReaderInterface>(
+    epoch: u64,
+    storage: T,
+) -> Result<LedgerInfoWithSignatures, Error> {
+    match storage.get_epoch_ending_ledger_infos(epoch, epoch) {
+        Ok(epoch_change_proof) => {
+            if let Some(ledger_info) = epoch_change_proof.ledger_info_with_sigs.get(0) {
+                Ok(ledger_info.clone())
+            } else {
+                Err(Error::StorageErrorEncountered(
+                    "Empty change proof found!".into(),
+                ))
+            }
+        }
+        Err(error) => Err(Error::StorageErrorEncountered(format!(
+            "Failed to get epoch ending ledger info! Error: {:?}",
+            error
+        ))),
+    }
+}
+
+/// Notifies a subscriber of new data according to the target ledger info
+fn notify_peer_of_new_data<T: StorageReaderInterface>(
+    cached_storage_server_summary: Arc<RwLock<StorageServerSummary>>,
+    config: StorageServiceConfig,
+    data_subscriptions: Arc<Mutex<HashMap<AccountAddress, DataSubscriptionRequest>>>,
+    lru_storage_cache: Arc<Mutex<LruCache<StorageServiceRequest, StorageServiceResponse>>>,
```

Comment:
not related to this PR, but I find it weird we need to pass both the cache and storage around, I imagine we should have an interface both cache and raw storage implements so it's swappable